include/ignition/rendering/BoundingBoxCamera.hh

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#ifndef IGNITION_RENDERING_BOUNDINGBOXCAMERA_HH_
+#define IGNITION_RENDERING_BOUNDINGBOXCAMERA_HH_
+
+#include <cstdint>
+#include <vector>
+
+#include <ignition/common/Event.hh>
+#include <ignition/math/Color.hh>
+#include <ignition/math/Vector3.hh>
+
+#include "ignition/rendering/Camera.hh"
+
+namespace ignition
+{
+  namespace rendering
+  {
+    /// \brief BoundingBox types for Visible / Full 2D Boxes / 3D Boxes
+    enum class BoundingBoxType
+    {
+      /// 2D box that shows the full box of occluded objects
+      BBT_FULLBOX2D = 0,
+
+      /// 2D box that shows the visible part of the
+      /// occluded object
+      BBT_VISIBLEBOX2D = 1,
+
+      /// 3D oriented box
+      BBT_BOX3D = 2
+    };
+
+    /// \brief 2D or 3D Bounding box. It stores the
+    /// position / orientation / size info of the box and its label
+    struct BoundingBox
+    {
+      /// \brief Box type
+      ignition::rendering::BoundingBoxType type;
+
+      /// \brief Center of the box in pixel coord in 2D, and camera coord in 3D
+      ignition::math::Vector3d center;
+
+      /// \brief Size of the box (width, height, depth), depth = 0 in 2D boxes
+      ignition::math::Vector3d size;
+
+      /// \brief Orientation of the 3D box in camera coord.
+      /// The 2D boxes are axis aligned (orientation = 0)
+      ignition::math::Quaterniond orientation;
+
+      /// \brief Label of the annotated object inside the box
+      uint32_t label;
```

Comment:
Just keep in mind that adding the members directly here, we won't be able to extend it in the future without breaking ABI. Using the PIMPL pattern would solve the issue.

Reply:
Is PIMPL used in `ign-rendering` since it has abstract interfaces in `include/ignition/rendering` and base implementations in `include/ignition/rendering/base`? I don't often see classes implemented in the `src` directory. But, I think using PIMPL is a good idea if we'd like to give ourselves flexibility down the road

Reply:
See the [HeightmapDescriptor](https://github.com/ignitionrobotics/ign-rendering/blob/ign-rendering5/include/ignition/rendering/HeightmapDescriptor.hh) for an example. Let me know if I can help with anything

Reply:
Done in a1103b52011a1545dc202370be90cd806169a55e